### PR TITLE
Update `Mode.MD_JSON` and add `Provider.DATABRICKS`

### PR DESCRIPTION
**Pull request by arcaputo3**
Databricks has added [support for the `openai` client](https://learn.microsoft.com/en-us/azure/databricks/machine-learning/model-serving/score-foundation-models#openaiclient-1) for foundational models such as DBRX and Llama 3 70B. This PR allows for integration with this endpoint by bypassing the `response_format` argument as this is not yet supported. Initial testing seems to work well.

Running the below notebook in Databricks
```python
import instructor
from openai import OpenAI
from pydantic import BaseModel


# Define your desired output structure
class UserInfo(BaseModel):
    name: str
    age: int


client = instructor.from_openai(
    OpenAI(
        api_key=dbutils.secrets.get("databricks", "token"),
        base_url=f"{dbutils.secrets.get('databricks', 'host')}/serving-endpoints",
    ),
    mode=instructor.Mode.DATABRICKS_JSON
)

user = client.chat.completions.create(
    messages=[
        {"role": "user", "content": "John Doe is 30 years old."},
    ],
    response_model=UserInfo,
    model="databricks-meta-llama-3-70b-instruct",
)
``` 
yields `UserInfo(name='John Doe', age=30)`.

TL;DR
- Treat similarly to other JSON Modes
- Skip adding a `response_format` (not yet supported by Databricks)
- Add provider syntax for explicit assertion

----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit f60d6250b245c366e97bcd38fa05f560110a0a4d  | 
|--------|

### Summary:
This PR introduces `Mode.DATABRICKS_JSON` to support JSON responses from Databricks, updating relevant enums and methods for integration.

**Key points**:
- Added `Mode.DATABRICKS_JSON` for handling Databricks JSON responses.
- Updated `Provider` enum to include `DATABRICKS`.
- Modified `from_openai` in `instructor/client.py` to assert `DATABRICKS_JSON` mode for Databricks provider.
- Updated `from_response` in `instructor/function_calls.py` to parse `DATABRICKS_JSON` mode.
- Added handling for `DATABRICKS_JSON` in `instructor/process_response.py`.
- Provided example usage in a Databricks notebook.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Review by jxnl — CHANGES_REQUESTED** [review]
please document this in patch.md

**Review by jxnl — COMMENTED** [review]
this also seems like noop, why wouldn't MD_JSON work with databricks as is

**Comment by arcaputo3** [comment]
> this also seems like noop, why wouldn't MD_JSON work with databricks as is

Had tried this originally, but Databricks seems to complain about two user messages in a row:

```
BadRequestError: Error code: 400 - {'error_code': 'BAD_REQUEST', 'message': 'Bad request: rpc error: code = InvalidArgument desc = Chat message input roles must alternate (user -> assistant -> u -> a -> ...) with an optional system at the start. Tool messages are optional and must follow a preceding assistant message containing tool calls.\n'}
File <command-3027138104257685>, line 20
      9     age: int
     12 client = instructor.from_openai(
     13     OpenAI(
     14         api_key=dbutils.secrets.get("databricks", "token"),
   (...)
     17     mode=instructor.Mode.MD_JSON
     18 )
---> 20 user = client.chat.completions.create(
     21     messages=[
     22         {"role": "user", "content": "John Doe is 30 years old."},
     23     ],
     24     response_model=UserInfo,
     25     model="databricks-meta-llama-3-70b-instruct",
     26 )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/instructor/client.py:91, in Instructor.create(self, response_model, messages, max_retries, validation_context, strict, **kwargs)
     80 def create(
     81     self,
     82     response_model: type[T],
   (...)
     87     **kwargs: Any,
     88 ) -> T | Awaitable[T]:
     89     kwargs = self.handle_kwargs(kwargs)
---> 91     return self.create_fn(
     92         response_model=response_model,
     93         messages=messages,
     94         max_retries=max_retries,
     95         validation_context=validation_context,
     96         strict=strict,
     97         **kwargs,
     98     )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/instructor/patch.py:143, in patch.<locals>.new_create_sync(response_model, validation_context, max_retries, strict, *args, **kwargs)
    131 @wraps(func)
    132 def new_create_sync(
    133     response_model: type[T_Model] = None,
   (...)
    138     **kwargs: T_ParamSpec.kwargs,
    139 ) -> T_Model:
    140     response_model, new_kwargs = handle_response_model(
    141         response_model=response_model, mode=mode, **kwargs
    142     )
--> 143     response = retry_sync(
    144         func=func,
    145         response_model=response_model,
    146         validation_context=validation_context,
    147         max_retries=max_retries,
    148         args=args,
    149         strict=strict,
    150         kwargs=new_kwargs,
    151         mode=mode,
    152     )
    153     return response
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/instructor/retry.py:152, in retry_sync(func, response_model, validation_context, args, kwargs, max_retries, strict, mode)
    149     raise ValueError("max_retries must be an int or a `tenacity.Retrying` object")
    151 try:
--> 152     for attempt in max_retries:
    153         with attempt:
    154             try:
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/tenacity/__init__.py:435, in BaseRetrying.__iter__(self)
    433 retry_state = RetryCallState(self, fn=None, args=(), kwargs={})
    434 while True:
--> 435     do = self.iter(retry_state=retry_state)
    436     if isinstance(do, DoAttempt):
    437         yield AttemptManager(retry_state=retry_state)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/tenacity/__init__.py:368, in BaseRetrying.iter(self, retry_state)
    366 result = None
    367 for action in self.iter_state.actions:
--> 368     result = action(retry_state)
    369 return result
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/tenacity/__init__.py:410, in BaseRetrying._post_stop_check_actions.<locals>.exc_check(rs)
    408 retry_exc = self.retry_error_cls(fut)
    409 if self.reraise:
--> 410     raise retry_exc.reraise()
    411 raise retry_exc from fut.exception()
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/tenacity/__init__.py:183, in RetryError.reraise(self)
    181 def reraise(self) -> t.NoReturn:
    182     if self.last_attempt.failed:
--> 183         raise self.last_attempt.result()
    184     raise self
File /usr/lib/python3.10/concurrent/futures/_base.py:451, in Future.result(self, timeout)
    449     raise CancelledError()
    450 elif self._state == FINISHED:
--> 451     return self.__get_result()
    453 self._condition.wait(timeout)
    455 if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:
File /usr/lib/python3.10/concurrent/futures/_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/instructor/retry.py:155, in retry_sync(func, response_model, validation_context, args, kwargs, max_retries, strict, mode)
    153 with attempt:
    154     try:
--> 155         response = func(*args, **kwargs)
    156         stream = kwargs.get("stream", False)
    157         response = update_total_usage(response, total_usage)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_utils/_utils.py:277, in required_args.<locals>.inner.<locals>.wrapper(*args, **kwargs)
    275             msg = f"Missing required argument: {quote(missing[0])}"
    276     raise TypeError(msg)
--> 277 return func(*args, **kwargs)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/resources/chat/completions.py:590, in Completions.create(self, messages, model, frequency_penalty, function_call, functions, logit_bias, logprobs, max_tokens, n, presence_penalty, response_format, seed, stop, stream, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)
    558 @required_args(["messages", "model"], ["messages", "model", "stream"])
    559 def create(
    560     self,
   (...)
    588     timeout: float | httpx.Timeout | None | NotGiven = NOT_GIVEN,
    589 ) -> ChatCompletion | Stream[ChatCompletionChunk]:
--> 590     return self._post(
    591         "/chat/completions",
    592         body=maybe_transform(
    593             {
    594                 "messages": messages,
    595                 "model": model,
    596                 "frequency_penalty": frequency_penalty,
    597                 "function_call": function_call,
    598                 "functions": functions,
    599                 "logit_bias": logit_bias,
    600                 "logprobs": logprobs,
    601                 "max_tokens": max_tokens,
    602                 "n": n,
    603                 "presence_penalty": presence_penalty,
    604                 "response_format": response_format,
    605                 "seed": seed,
    606                 "stop": stop,
    607                 "stream": stream,
    608                 "stream_options": stream_options,
    609                 "temperature": temperature,
    610                 "tool_choice": tool_choice,
    611                 "tools": tools,
    612                 "top_logprobs": top_logprobs,
    613                 "top_p": top_p,
    614                 "user": user,
    615             },
    616             completion_create_params.CompletionCreateParams,
    617         ),
    618         options=make_request_options(
    619             extra_headers=extra_headers, extra_query=extra_query, extra_body=extra_body, timeout=timeout
    620         ),
    621         cast_to=ChatCompletion,
    622         stream=stream or False,
    623         stream_cls=Stream[ChatCompletionChunk],
    624     )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_base_client.py:1240, in SyncAPIClient.post(self, path, cast_to, body, options, files, stream, stream_cls)
   1226 def post(
   1227     self,
   1228     path: str,
   (...)
   1235     stream_cls: type[_StreamT] | None = None,
   1236 ) -> ResponseT | _StreamT:
   1237     opts = FinalRequestOptions.construct(
   1238         method="post", url=path, json_data=body, files=to_httpx_files(files), **options
   1239     )
-> 1240     return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_base_client.py:921, in SyncAPIClient.request(self, cast_to, options, remaining_retries, stream, stream_cls)
    912 def request(
    913     self,
    914     cast_to: Type[ResponseT],
   (...)
    919     stream_cls: type[_StreamT] | None = None,
    920 ) -> ResponseT | _StreamT:
--> 921     return self._request(
    922         cast_to=cast_to,
    923         options=options,
    924         stream=stream,
    925         stream_cls=stream_cls,
    926         remaining_retries=remaining_retries,
    927     )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_base_client.py:1005, in SyncAPIClient._request(self, cast_to, options, remaining_retries, stream, stream_cls)
   1003 if retries > 0 and self._should_retry(err.response):
   1004     err.response.close()
-> 1005     return self._retry_request(
   1006         options,
   1007         cast_to,
   1008         retries,
   1009         err.response.headers,
   1010         stream=stream,
   1011         stream_cls=stream_cls,
   1012     )
   1014 # If the response is streamed then we need to explicitly read the response
   1015 # to completion before attempting to access the response text.
   1016 if not err.response.is_closed:
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_base_client.py:1053, in SyncAPIClient._retry_request(self, options, cast_to, remaining_retries, response_headers, stream, stream_cls)
   1049 # In a synchronous context we are blocking the entire thread. Up to the library user to run the client in a
   1050 # different thread if necessary.
   1051 time.sleep(timeout)
-> 1053 return self._request(
   1054     options=options,
   1055     cast_to=cast_to,
   1056     remaining_retries=remaining,
   1057     stream=stream,
   1058     stream_cls=stream_cls,
   1059 )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-69af8d21-258f-490b-9aa2-c053792a0ead/lib/python3.10/site-packages/openai/_base_client.py:1020, in SyncAPIClient._request(self, cast_to, options, remaining_retries, stream, stream_cls)
   1017         err.response.read()
   1019     log.debug("Re-raising status error")
-> 1020     raise self._make_status_error_from_response(err.response) from None
   1022 return self._process_response(
   1023     cast_to=cast_to,
   1024     options=options,
   (...)
   1027     stream_cls=stream_cls,
   1028 )
   ```

**Comment by jxnl** [comment]
i see in anthropic we merge them, maybe we should just avoid two user messages in a row, would you be willing to make that change?

**Comment by arcaputo3** [comment]
> i see in anthropic we merge them, maybe we should just avoid two user messages in a row, would you be willing to make that change? i think we have a merge unction that we use in anthropic

Definitely, so basically we just want to merge the last user message instead of appending within `MD_JSON`? Happy to add that change and seems better from a generalizability perspective.

File: docs/concepts/patching.md [hunk 1]
```diff
@@ -54,15 +54,31 @@ client = instructor.from_openai(OpenAI(), mode=instructor.Mode.JSON)
...
+
+Databricks syntax:
+```python
+import instructor
+from openai import OpenAI
+
+# Assuming Databricks secrets are set
+client = instructor.from_openai(
+    OpenAI(
+        api_key=dbutils.secrets.get("databricks", "token"),
```
Comment (jxnl) [inline docs/concepts/patching.md:79]: is dbutils a you thing? i imagine this is just env variables right?

**Review by jxnl — APPROVED** [review]
